### PR TITLE
Added user schema for db

### LINKED ISSUE #157: Adding a model/schema for the "users" collection in the database
**Is your feature request related to a problem? Please describe.**
As we move forward to creating support for users. We should first set up a model in the database. 

**Describe the solution you'd like**
Follow the schema diagram on the wiki to create `users.js` schema under `se701-assignment1/server/src/db/models`

### PR DESCRIPTION
**Pull request by RajanGuptaNZ**
Fixes #157 

## Proposed Changes

  - Add users schema to db according to the schema diagram on the wiki
  - Add test users to db until endpoints implemented properly
  - Set up endpoint routing and implement basic 'add user' endpoint (needs work but basic functionality works)

### CODE REVIEW

File: server/src/db/index.js [hunk 1]
```diff
@@ -1,5 +1,8 @@
 const mongoose = require('mongoose');
 
+// Removes deprecation warning when enforcing uniquness
+mongoose.set('useCreateIndex', true);
```
Comment (MingChen91) [inline server/src/db/index.js:4]: I think the warning actually existed already and is not just for the uniqueness thing, I'm running current master and it shows the same deprecation warning. I'm actually not sure if it's an issue, but this does get rid of the warning. Anyone else have any ideas?

File: frontend/.directory [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[Dolphin]
+Timestamp=2020,4,6,12,21,29
+Version=4
+
+[Settings]
+HiddenFilesShown=true
```
Comment (MingChen91) [inline frontend/.directory:6]: Is this something that should be in .gitignore?

File: server/.directory [hunk 3]
```diff
@@ -0,0 +1,6 @@
+[Dolphin]
+Timestamp=2020,4,6,11,39,35
+Version=4
+
+[Settings]
+HiddenFilesShown=true
```
Comment (MingChen91) [inline server/.directory:6]: Is this something that should be in .gitignore?

File: server/src/db/index.js (hunk 1 above)
Comment (RajanGuptaNZ) [reply server/src/db/index.js:4]: Warning would disappear when I removed the uniqueness check. But I was working on the version before #161 was merged. Jacob has added a uniqueness check too which is why the warning will still appear

File: frontend/.directory (hunk 2 above)
Comment (RajanGuptaNZ) [reply frontend/.directory:6]: Yep, don't know how these got included, I'll fix asap

File: server/src/db/models/users.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+const userSchema = new mongoose.Schema({
+  email: {
+    type: String,
+    required: true,
+    unique: true,
+  },
+  posts: {
+    type: Array,
+    default: [],
+    required: true,
```
Comment (jacoballen4534) [inline server/src/db/models/users.js:12]: I don't think the 'required' is needed, as it gets a default if no posts are provided. Same for all following entries.

File: server/.directory (hunk 3 above)
Comment (RajanGuptaNZ) [reply server/.directory]: yep, removed

File: frontend/.directory (hunk 2 above)
Comment (RajanGuptaNZ) [reply frontend/.directory]: Removed

File: server/src/db/models/users.js (hunk 4 above)
Comment (RajanGuptaNZ) [reply server/src/db/models/users.js]: Yep removed & tested and it still works. So updated PR has these lines removed

File: server/src/db/models/users.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+const mongoose = require('mongoose');
+
+const userSchema = new mongoose.Schema({
+  email: {
```
Comment (jacoballen4534) [inline server/src/db/models/users.js:4]: Are we just using email, or also storing a token of some sorts?

**Comment by jacoballen4534** [comment]
> I think some of the files should be in gitignores, and there is a bunch of changes in package lock > where additional "optional:true" parameters were added, does anyone else know what that ? means?

I believe the  "optional:true" should be there, somehow package.lock got committed without it before.

File: server/src/routes/users.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  { email: 'user1@gmail.com', posts: [], comments: [], claps: [], laughs: [], sads: [] },
+  { email: 'user2@gmail.com', posts: [], comments: [], claps: [], laughs: [], sads: [] },
+  { email: 'user3@gmail.com', posts: [], comments: [], claps: [], laughs: [], sads: [] },
+  { email: 'user4@gmail.com', posts: [], comments: [], claps: [], laughs: [], sads: [] },
+  { email: 'user5@gmail.com', posts: [], comments: [], claps: [], laughs: [], sads: [] },
+];
+
+// Insert test users if document is empty
+// eslint-disable-next-line func-names
+User.countDocuments(function(err, count) {
```
[the remark this answers is not in the document]
Comment (RajanGuptaNZ) [reply server/src/routes/users.js]: Just tried this, can you check it again? I'm new to arrow functions so not 100% sure I did it right. Thanks!

File: server/src/routes/users.js [hunk 7]
```diff
@@ -14,15 +14,16 @@ const testUsers = [
...
-User.countDocuments(function(err, count) {
-  if (!err && count === 0) {
-    // eslint-disable-next-line func-names
-    User.insertMany(testUsers, function(err1) {
-      if (err1) throw err1;
-    });
-  }
-});
+insertTestUsers = () => {
+  User.countDocuments(function(err, count) {
```
Comment (jacoballen4534) [inline server/src/routes/users.js:18]: I was meaning this function. Eg
`User.countDocuments((err, count) => {`

File: server/src/db/index.js (hunk 1 above)
Comment (MingChen91) [reply server/src/db/index.js:4]: ah right. sweet

File: server/src/db/models/users.js (hunk 5 above)
Comment (MingChen91) [reply server/src/db/models/users.js:4]: I think email would be enough? @wshin387 is just email enough? like when then authenticate and the middleware says ok, can I just then take the email? or should we store the token somewhere too

File: server/src/routes/users.js (hunk 7 above)
Comment (RajanGuptaNZ) [reply server/src/routes/users.js]: Done, Thanks for your help!

**Pull request merged**